### PR TITLE
[DistillationTrainer refactor] Extract the teacher-server path into ServerDistillationTrainer

### PR DESCRIPTION
**Pull request by qgallouedec**
*Plan steps 1.1 + 1.2 (merged) — stacked on PR #6453 . Part of #6449.*

The server-backed path — fetch per-token teacher logprobs from a vLLM server instead of running a local teacher forward — is a distinct use case that only ever runs over a **sparse top-k support** (the server returns top-k logprobs, never the full distribution). Move it out of the base trainer into a dedicated experimental subclass so the base can converge on the local, full-vocabulary objective.

### New package `trl/experimental/server_distillation/`

- **`ServerDistillationTrainer(DistillationTrainer)`** overrides `compute_loss` with the server path and carries the moved `_get_teacher_token_logprobs_from_server`, `_compute_server_sparse_top_1_divergence_loss`, `_compute_server_forward_kl_loss`, and the module-level `build_teacher_request_inputs`. It sets up the `VLLMClient` in `__init__` after `super().__init__(teacher_model=None)`.
- **`ServerDistillationConfig(DistillationConfig)`** adds `teacher_model_server_url` and carries the server-only validations, now unconditional (the subclass *is* the server trainer, so the old `use_teacher_server` flag is gone).

The shared sparse helpers (`_compute_sparse_top_1_divergence_loss`, `_get_reverse_kl_top_1_tokens`, `_add_tail_bucket`, `_jsd_divergence`, `_reduce_divergence_loss`) are **inherited, not copied** — the base still uses them for its local top-1 path. They migrate to the subclass later, once the base local sparse path is removed (plan 1.5/1.6).

### Base trainer / config

- `__init__` drops the server branch; a `None` teacher just leaves `self.teacher_model = None`.
- `compute_loss` drops the server dispatch; the local path runs unconditionally. **5 dispatch paths → 2** (local top-1, local full-vocab) plus Liger.
- `_get_teacher_logits` drops the server `NotImplementedError`.
- config drops `use_teacher_server` / `teacher_model_server_url` and their `__post_init__` checks.

### Tests

Server tests move to `tests/experimental/test_server_distillation_trainer.py`, retargeted at `ServerDistillationTrainer`/`ServerDistillationConfig`.

The server-side infrastructure (`/get_sequence_logprobs/`, `VLLMClient`) is untouched and still used by SDFT/SDPO.

### Verification done without pytest

- Both packages import; `ServerDistillationTrainer` is a `DistillationTrainer` subclass.
- Base trainer/module no longer expose any server symbol; subclass has them and inherits the shared sparse helpers.
- All 4 config validations fire with the new messages; a valid config builds.
- The moved server loss functions were exercised on CPU with fabricated tensors: finite forward+backward on a ragged batch (`beta=1` and `beta=0`), and an identical student/teacher distribution gives exactly `0`.

### Risk

**High — needs runtime verification.** The full server `compute_loss` and the `__init__` client wiring are only exercised by the mocked-client tests, which were not run here. The numerical checks cover the loss math but not the end-to-end training step. Run `tests/experimental/test_server_distillation_trainer.py` and `tests/experimental/test_distillation_trainer.py` before merging.

### Note

This PR also drops an orphaned test (`test_distillation_config_rejects_invalid_distillation_objective`) — see PR #6453, which was amended to remove it at the point the field was deleted, so it does not resurface here.

---

> [!NOTE]
> **High Risk**
> Large refactor of training loss paths and a breaking removal of `use_teacher_server` on `DistillationConfig`; callers must migrate to `ServerDistillationTrainer` and run the distillation test suites before merge.
> 
> **Overview**
> **External teacher distillation is split into a dedicated experimental API.** vLLM server scoring (sparse top-k logprobs, ragged batches, padding sentinels) no longer lives behind `use_teacher_server` on the base trainer.
> 
> New **`ServerDistillationTrainer`** subclasses **`DistillationTrainer`**, wires **`VLLMClient`** in `__init__`, and owns `compute_loss` plus `build_teacher_request_inputs` and the server loss helpers. **`ServerDistillationConfig`** adds required **`teacher_model_server_url`** and keeps the former server-only validations (Liger, `reverse_kl_top_1_mode='argmax'`, `loss_top_k` vs `beta`).
> 
> **`DistillationTrainer` / `DistillationConfig`** are local-teacher only: server fields and dispatch are removed; `compute_loss` is always the local full-vocab or local top-1 path (plus Liger). Tokenizer mismatch guidance now points callers at **`ServerDistillationTrainer`**.
> 
> Server-focused tests move to **`tests/experimental/test_server_distillation_trainer.py`**; **`test_distillation_trainer.py`** keeps local distillation coverage. **`ServerDistillationTrainer`** is registered in **`base_trainer`** exports.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 9ad8c2a5610e0df01eb5d6c1d5e6d08fc6eccc37. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/server_distillation/server_distillation_trainer.py [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+                f"Invalid prompt length {prompt_len} for trimmed sequence of length {trimmed_row.numel()}."
+            )
+        trimmed_input_ids.append(trimmed_row.tolist())
+        prompt_lengths_list.append(prompt_len)
+        completion_lengths_list.append(int(trimmed_row.numel()) - prompt_len)
+
+    return trimmed_input_ids, prompt_lengths_list, completion_lengths_list
+
+
+class ServerDistillationTrainer(DistillationTrainer):
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/server_distillation/server_distillation_trainer.py:69]: Intentional here, this is experimental scaffolding, kept a thin subclass on purpose so the PR stays reviewable. It's re-homed onto the self-contained stable trainer later in the series (#6449, item 59). The self-contained mandate applies to the promoted trainer.

File: trl/experimental/server_distillation/server_distillation_trainer.py [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+        eval_dataset=None,
+        processing_class=None,
+        compute_metrics=None,
+        callbacks=None,
+        optimizers=(None, None),
+        preprocess_logits_for_metrics=None,
+        peft_config=None,
+    ):
+        if args is None:
+            args = ServerDistillationConfig(output_dir="tmp_server_distillation")
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/server_distillation/server_distillation_trainer.py:97]: the fallback is copied from the base trainer's pattern (consistency mandate)

File: trl/experimental/distillation/distillation_config.py [hunk 3]
```diff
@@ -79,20 +79,12 @@ class DistillationConfig(_BaseConfig):
         teacher_model_init_kwargs (`dict[str, Any]` or `None`, *optional*):
             Keyword arguments passed to `AutoModelForCausalLM.from_pretrained` when instantiating the teacher model
             from a string.
-        use_teacher_server (`bool`, *optional*, defaults to `False`):
-            Whether to use an external vLLM teacher server instead of a local teacher model.
-        teacher_model_server_url (`str` or `None`, *optional*):
-            Base URL of a vLLM server hosting the teacher model (e.g., `"http://localhost:8000"`). When set, teacher
-            logprobs are fetched from the server instead of running a local forward pass when `use_teacher_server=True`.
```
Comment (cmpatino) [inline trl/experimental/distillation/distillation_config.py:86]: We should also remove these from distillation_trainer.md

**Review by cmpatino — COMMENTED** [review]
Do you have specific reasons for avoiding support for external teacher servers in the refactored trainer?

I think only supporting local teachers limits the spectrum of the distillation runs the trainer can support. I'm aware it introduces a lot of complexity in the trainer, but it's maybe something we can look into instead of entirely removing the capability.

One thing that would significantly simplify the logic is using `vllm>=0.20` to retrieve specific logprob IDs from the server.

But open to discussion if we want to ship the trainer without the server support and focus on other features.

**Comment by cmpatino** [comment]
I take back my earlier comment about trying to keep the external teacher server for this trainer. See full comment here https://github.com/huggingface/trl/pull/6456#issuecomment-5023582170
